--- a/contrib/imczmq/imczmq.c
+++ b/contrib/imczmq/imczmq.c
@@ -308,7 +308,6 @@ static rsRetVal rcvData(){
 		pData = zlist_next(listenerList);
 	}
 
-	zframe_t *frame;
 	zsock_t *which = (zsock_t *)zpoller_wait(poller, -1);
 	while(which) {
 		if (zpoller_terminated(poller)) {
@@ -323,8 +322,13 @@ static rsRetVal rcvData(){
 			DBGPRINTF("imczmq: found matching socket\n");
 		}
 
-		frame = zframe_recv(which);
-		char *buf = zframe_strdup(frame);
+		zframe_t *frame = zframe_recv(which);
+		char *buf = NULL;
+
+		if (frame != NULL)
+			buf = zframe_strdup(frame);
+
+		zframe_destroy(&frame);
 
 		if(buf == NULL) {
 			DBGPRINTF("imczmq: null buffer\n");
@@ -348,7 +352,6 @@ static rsRetVal rcvData(){
 		which = (zsock_t *)zpoller_wait(poller, -1);
 	}
 finalize_it:
-	zframe_destroy(&frame);
 	zpoller_destroy(&poller);
 	pData = zlist_first(listenerList);
 	while(pData) {
