---
 tools/omfwd.c |   23 +++++++++++++++--------
 1 file changed, 15 insertions(+), 8 deletions(-)

--- a/tools/omfwd.c
+++ b/tools/omfwd.c
@@ -488,14 +488,21 @@
 		} else {
 			dbgprintf("error forwarding via udp, suspending\n");
 			if(pWrkrData->errsToReport > 0) {
-				errmsg.LogError(lasterrno, RS_RET_ERR_UDPSEND,
-						"omfwd: error %d sending "
-						"via udp", lasterrno);
-				if(pWrkrData->errsToReport == 1) {
-					errmsg.LogMsg(0, RS_RET_LAST_ERRREPORT, LOG_WARNING, "omfwd: "
-							"max number of error message emitted "
-							"- further messages will be "
-							"suppressed");
+				/* Do not report EINVAL errors, as these are transient at start-up
+				 * when a source address is configured, even with Ipfreebind=1.
+				 * The omission of this error is only suited to systems where the
+				 * configured source address is guaranteed to be valid.
+				 */
+				if (lasterrno != EINVAL) {
+					errmsg.LogError(lasterrno, RS_RET_ERR_UDPSEND,
+							"omfwd: error %d sending "
+							"via udp", lasterrno);
+					if(pWrkrData->errsToReport == 1) {
+						errmsg.LogMsg(0, RS_RET_LAST_ERRREPORT, LOG_WARNING, "omfwd: "
+							      "max number of error message emitted "
+							      "- further messages will be "
+							      "suppressed");
+					}
 				}
 				--pWrkrData->errsToReport;
 			}
